File: api/transcript.py
from http.server import BaseHTTPRequestHandler
import json
import re
import urllib.request
from urllib.parse import urlparse, parse_qs

UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120 Safari/537.36"


def _parse_vtt(vtt: str) -> str:
    texts = []
    for line in vtt.split("\n"):
        line = line.strip()
        if not line or "-->" in line or line.startswith("WEBVTT") or re.match(r"^\d+$", line):
            continue
        text = re.sub(r"<[^>]+>", "", line).strip()
        if text and (not texts or texts[-1] != text):
            texts.append(text)
    return " ".join(texts)


def _fetch(video_id: str) -> str:
    req = urllib.request.Request(
        f"https://www.youtube.com/watch?v={video_id}&hl=en",
        headers={"User-Agent": UA},
    )
    with urllib.request.urlopen(req, timeout=15) as r:
        html = r.read().decode("utf-8", errors="ignore")

    m = re.search(r'"captionTracks":\s*(\[.*?\])', html)
    if not m:
        raise ValueError("자막 없음")

    tracks = json.loads(m.group(1))
    track = next((t for t in tracks if t.get("languageCode", "").startswith("en")), None)
    if not track:
        track = tracks[0] if tracks else None
    if not track or not track.get("baseUrl"):
        raise ValueError("자막 URL 없음")

    vtt_req = urllib.request.Request(track["baseUrl"] + "&fmt=vtt", headers={"User-Agent": UA})
    with urllib.request.urlopen(vtt_req, timeout=15) as r:
        vtt = r.read().decode("utf-8", errors="ignore")

    return _parse_vtt(vtt)


class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        params = parse_qs(urlparse(self.path).query)
        video_id = (params.get("v") or [None])[0]

        if not video_id:
            self._respond(400, {"error": "Missing v parameter"})
            return
        try:
            transcript = _fetch(video_id)
            self._respond(200, {"transcript": transcript, "video_id": video_id})
        except Exception as e:
            self._respond(500, {"error": str(e)})

    def _respond(self, code: int, body: dict):
        data = json.dumps(body, ensure_ascii=False).encode()
        self.send_response(code)
        self.send_header("Content-Type", "application/json")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(data)

    def log_message(self, *args):
        pass

File: api/transcript.js
export default async function handler(req, res) {
  res.setHeader("Access-Control-Allow-Origin", "*");
  const videoId = req.query.v;
  if (!videoId) return res.json({ status: "ok" });

  const errors = [];

  // 시도 1: TV embedded 클라이언트
  try {
    const t = await fetchInnertube(videoId, "TVHTML5_SIMPLY_EMBEDDED_PLAYER", "2.0");
    return res.json({ transcript: t, video_id: videoId });
  } catch (e) { errors.push("TV:" + e.message); }

  // 시도 2: iOS 클라이언트
  try {
    const t = await fetchInnertube(videoId, "IOS", "19.09.3");
    return res.json({ transcript: t, video_id: videoId });
  } catch (e) { errors.push("iOS:" + e.message); }

  // 시도 3: timedtext 직접 접근
  try {
    const t = await fetchTimedtext(videoId);
    return res.json({ transcript: t, video_id: videoId });
  } catch (e) { errors.push("timedtext:" + e.message); }

  res.status(500).json({ error: errors.join(" | ") });
}

async function fetchInnertube(videoId, clientName, clientVersion) {
  const resp = await fetch("https://www.youtube.com/youtubei/v1/player", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      videoId,
      context: { client: { clientName, clientVersion, hl: "en" } },
    }),
  });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
  const data = await resp.json();
  const tracks = data?.captions?.playerCaptionsTracklistRenderer?.captionTracks;
  if (!tracks?.length) throw new Error("자막 없음");
  const track = tracks.find(t => t.languageCode?.startsWith("en")) || tracks[0];
  if (!track?.baseUrl) throw new Error("URL 없음");
  const vtt = await (await fetch(track.baseUrl + "&fmt=vtt")).text();
  return parseVtt(vtt);
}

async function fetchTimedtext(videoId) {
  // 자막 목록 먼저 가져오기
  const listResp = await fetch(
    `https://www.youtube.com/api/timedtext?v=${videoId}&type=list&fmt=json3`
  );
  if (!listResp.ok) throw new Error(`HTTP ${listResp.status}`);
  const list = await listResp.json();
  const tracks = list?.captions?.playerCaptionsTracklistRenderer?.captionTracks;
  if (!tracks?.length) throw new Error("자막 없음");
  const track = tracks.find(t => t.languageCode?.startsWith("en")) || tracks[0];
  const vtt = await (await fetch(track.baseUrl + "&fmt=vtt")).text();
  return parseVtt(vtt);
}

function parseVtt(vtt) {
  const texts = [];
  for (const raw of vtt.split("\n")) {
    const line = raw.trim();
    if (!line || line.includes("-->") || line.startsWith("WEBVTT") || /^\d+$/.test(line)) continue;
    const text = line.replace(/<[^>]+>/g, "").trim();
    if (text && texts[texts.length - 1] !== text) texts.push(text);
  }
  return texts.join(" ");
}
